### ISSUE TITLE
Node-local API cache

### ISSUE THREAD

**Issue by tallclair** [kubernetes/kubernetes#84248]
**What would you like to be added**:

A dedicated node-local API for retrieving node-local resources, including:
- Pods scheduled to the node
- The node object itself

The API should have clear semantics for adding additional resources as needed. This includes consistent API paths, SubjectAccessReview patterns, SLOs around cache accuracy, and cache-miss handling.

**Why is this needed**:

For large-scale clusters, an increasing number of Daemons are depending on the nodes `/pods` API. This API has a number of issues, including:
- Using the authenticated port, it share's a permission with the root-on-node `node/proxy` permission (see https://github.com/kubernetes/kubernetes/issues/83465)
- The API uses a different source of truth from the API server, and doesn't have any consistency guarantees.
- The API doesn't conform to Kubernetes API standards.

**Open design questions**:

- Should the cache be maintained & served by the Kubelet? Or a DaemonSet?
  - If the latter, should the Kubelet read from the local cache?
- If I can GET a specific through the k8s API, I should also be able to get it from the local cache. But its also convenient to be able to say "you can only get pods running on the same node". Do we need to support both use cases? How should this be structured?
  - Idea: the API could come in 2 "flavors", one would look exactly like the cluster API, and fallback to the cluster API on a cache-miss (list requests need to be handled differently). The second flavor would be the local-only version, that uses different SAR attributes, and has a different request path (e.g. prefix `local-api/`)
- Should the local API support full apiserver features? Including watch, filter, etc.? If so, it probably makes sense to reuse the apimachinery.

/sig node
/sig scalability
/sig apimachinery
/sig auth
/sig instrumentation

/priority important-longterm
/help

**Comment by tallclair** [comment]
related: https://github.com/kubernetes/kubernetes/issues/28610

**Comment by draveness** [comment]
Hi @tallclair, do you mind if start to work on this?

**Comment by draveness** [comment]
Per offline discussion with Tim. This is a very big issue. And I'm going to work on it. Thanks @gongguan 

/assign

**Comment by gongguan** [comment]
> Per offline discussion with Tim. This is a very big issue. And I'm going to work on it. Thanks @gongguan
> 
> /assign

Ok, I'll watch and learn.

**Comment by tedyu** [comment]
It would be good to get agreement on the points raised under Open design questions before implementation starts.

**Comment by draveness** [comment]
Yes, I'll open a KEP first before pulling the implementation

Regards,
Draven

**Comment by tallclair** [comment]
Notes from sig-node discussion 2019-10-29:

- Question: If we build an apimachinery API on the node, would we want to include other (not just RO APIs) node APIs in the new API? Or scope it just to a local cache?
- Concern: Running more software on the node multiplies the resource costs by the number of nodes.
  - Alternative: Run a caching server for a group of nodes, rather than for each node
- Concern: How much can different use cases tolerate staleness of data? E.g. implications for mounting a persistent volume if it's not unmounted from another node.
- **Action Item:** Start with concrete use cases. What are the pain points with the current approach? What are the requirements for the proposed API?

**Comment by lavalamp** [comment]
I'm concerned by the security implications of a second entrypoint into cluster data. I don't really see how to shoehorn an apiserver into kubelet in a way that solves the potential inconsistency of authn/authz rules being applied by different components; additionally, the need to run such rules means that a cache can never really be truly local.

Is there a reason we can't solve this by "just" scaling kube-apiserver really big? (i.e., horizontally, by making many replicas, which was the original design.) And/or operating a Deployment of apiserver write-through caches (which is much more work for a very similar outcome).

**Comment by khenidak** [comment]
It might be worth doing the #s exercise comparing having kubelet as cache on every node, and the implication on its min h/w requirements etc vs. either by _scaling api-server as @lavalamp said_ or _cache(write through or not) as a replicated stateless pod_  presenting the same api endpoint as apiserver and acts as as a target for `kubernetes default service`, and can only be deployed if the cluster operator needs it.

Also some (I don't have an actual #) daemonset instances are more concerned with data beyond what is relevant to kubelet (and what kubelet will cache), or the node it sets on. daemonsets that uses api server for coordinations will be probably interested in CRDs' instances that are not directly linked to `current node`, do we expect clients to still call apiserver for these?
   
kal

**Comment by lavalamp** [comment]
> do we expect clients to still call apiserver for these?

I think folks will find an apiserver cache with incomplete resource support very confusing.

**Comment by deads2k** [comment]
> > do we expect clients to still call apiserver for these?
> 
> I think folks will find an apiserver cache with incomplete resource support very confusing.

Our authn/authz already uses caches to validate SA tokens and RBAC rules.  Barring some latency races, it hasn't been much of a practical issue.  

By allowing the kubelet to server an API surface that is compatible with the kube-apiserver, but which only contains those resources that are present on the node, we can make it easier to develop more narrowly scoped plugins. 

I would not start by allowing writing back to the kube-apiserver, but a view of pods, configmaps, secrets, PVs, and PVCs that are being used by the present kubelet seems valuable.  My only real concern about such a design is how we would allow projection of custom resources down.  I think we would quickly end up in cases where people wanted to project their custom resources for CNI or CSI down via this mechanism.  Also, a pods/logs endpoint would be pretty handy.

**Comment by serathius** [comment]
tldr; For now I would propose for kubelet to support "node" and "pods" resources but with full benefits of k8s apimachinery.

Idea for node-local cache was born from rising interest in building node base logging and monitoring agents that can scale. Instead of developing autoscaling and sharding for cluster wide applications, we can deploy per node agents running in DaemonSet that scale horizontally with the cluster.  

Node based agents (e.g. Fluend for logging) require access to metadata to observed resources local to node (pods/node). Fetching this metadata from apiserver directly also creates a scalability problem. Even when those agents correctly fetch pods filtering from single node, apiserver still does all filtering in memory.

Node agents are mainly interested in two resources Pods running on the node, and Node object itself. "/pods" was implemented long ago and there is work on [proposal for "/node" proposed](https://github.com/kubernetes/enhancements/pull/1038). Main con of currently exposed endpoints is that they are not K8s API but plain JSON. This means that agents are only restricted to periodically scrape kubelet and do very expensive parsing of JSON. Maybe parsing of json object describing tens of containers doesn't seem costly but for 5k clusters adding 10MBs per node results in allocating additional 50GB of memory just for monitoring pipeline. 

If SIG-node wants to go in direction of proxying and caching requests to apiserver, then I think migrating APIs to apimachinery would be very beneficial and be backed up by real needs.

**Comment by tallclair** [comment]
I agree with restricting this to read-only pod & node requests for now, and only expanding it on a case-by-case basis addressing real user requirements. I don't see much benefit to making it write-through.

My primary motivation is to remove the /pods endpoint, which is currently poorly supported. I agree that horizontally scaling the apiserver or a cluster-wide write-through cache could address this, but I can't speak to the tradeoffs. My original proposal _was_ to delete /pods, and send requests directly to the apiserver (scaling as needed). I'll let @serathius or someone from @kubernetes/sig-scalability-feature-requests  comment on that tradeoff.

**Comment by tedyu** [comment]
@serathius 
Can you chime in w.r.t. @tallclair 's comment above ?

Thanks

**Comment by jacksontj** [comment]
As a user of the `/pods` endpoint -- the main advantage I have by using it is that I don't rely on the control plane availability to get the data. This is actually very important from a reliability perspective (lots of details in my [katalog-sync post](https://medium.com/wish-engineering/katalog-sync-reliable-integration-of-consul-and-kubernetes-ebe8aae0852a) if you are interested). But suffice it to say that a node-local API is a good thing to have, assuming proper permissions etc.

That being said, I do wish (as asked in #85315) that the pods endpoint on the node supported watching -- as right now I have to poll that endpoint which is definitely sub-optimal.

**Comment by clamoriniere** [comment]
Hi,

I created the [`/node` endpoint KEP](https://github.com/kubernetes/enhancements/pull/1038), and I think it has some synergy with this issue.

I would like to add some comments on @serathius remark. I think that the order to magnitude mentioned is a bit overestimated. I check on several nodes and the current '/pods' JSON payload is about 80ko for 10 pods; the maximum number of pod per Kubelet recommended is 110 (What we usually see it is node running less than 20 pods). So If I do some simple math, the maximum '/pods' payload size is normally around 880ko, so let's say < 1mo.

Also since '/pods' return only the list of pods managed by the Kubelet himself, it will never exceed 1mo event if the cluster has 5k nodes. This is one of the reasons we use this endpoint with the Datadog-agent to retrieve pods information.

If we follow the same logic and extend the Kubelet endpoints with `/node`, the new endpoint will return only one Node information that corresponds to the Kubelet’s node. So the `/node` payload will be also limited to around 10ko (I looked on a large 2.5k cluster and the average is 7.8ko)


@tallclair  if I understand your goal is to replace the `/pods` by a new better maintained Kubelet endpoint (RBAC, versioning, protobuf/json support...). But I wanted to have the confirmation that the new endpoint still returns only “local” Pods (running on the same node). Is that correct?

**Comment by tallclair** [comment]
Yes, definitely - only node & pods scheduled to the node. I would be totally fine with scoping this to ONLY pods & node, but if we did extend it to other resources I would expect the supported resources to be similar to the graph that the node authorizer maintains.

**Comment by lukasmrtvy** [comment]
What about labels ? Related: https://github.com/kubernetes/kubernetes/issues/80745

**Comment by tallclair** [comment]
I still think this is an important issue for large scale clusters, which continue to rely on the underspecified `/pods` endpoint on the kubelet for this.

**Comment by dims** [comment]
cc @adisky

**Comment by dims** [comment]
long-term-issue (note to self)

**Comment by smarterclayton** [comment]
I feel like this is the wrong direction (something specific to one off pods and things scoped to a resource).  A node local cache should survive disruption to the apiserver - if everyone has to do different things for extension resources, that means you only get availability for some of your node local workloads.  I agree this needs a KEP, but solving for a very narrow set of resources is solving for overly simplistic environments.

**Comment by tallclair** [comment]
/remove-lifecycle stale
/lifecycle frozen

This needs a lot more discussion, but could have a lot of value.

**Comment by tallclair** [comment]
This needs a KEP to progress. But even before writing a KEP for this feature, we need to decide what the future intention for the Kubelet API is.

**Comment by clamoriniere** [comment]
Hi @tallclair 

Do you know where the discussion about "future intention for the Kubelet API" has already started somewhere?
If not, how can we start it?
